zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,527 @@
+    input: Column<Advice>, // input bitstream byte/row
+    // output parts:
+    // encoded value
+    data: Column<Advice>,
+    data_rlc: Column<Advice>,
+    // encoded bit length
+    num_bits: Column<Advice>,
+    // randomness multiplier selector
+    r_multi: Column<Advice>,
+    // the factor with which the part needs to be multiplied
+    shift_factor: Column<Advice>,
+    // the factor with which the part needs to be multiplied when
+    // we decode it
+    decode_factor: Column<Advice>,
+    // fixed huffman table
+    huffman_table: [TableColumn; 3], // [value, code, code_length]
+    fixed_table: [TableColumn; 2],   // [tag, or(exponent(2**0..8), exponent(2**0..32), bool(0,1))]
+
+    _marker: PhantomData<F>,
+}
+
+/// Circuit configuration arguments
+pub struct CompressCircuitConfigArgs<F> {
+    // randomness used to do RLC
+    pub randomness: Expression<F>,
+}
+
+impl<F: Field> SubCircuitConfig<F> for CompressCircuitConfig<F> {
+    type ConfigArgs = CompressCircuitConfigArgs<F>;
+
+    fn new(
+        meta: &mut ConstraintSystem<F>,
+        CompressCircuitConfigArgs { randomness }: Self::ConfigArgs,
+    ) -> Self {
+        let q_enable = meta.complex_selector();
+        let q_end = meta.selector();
+        let q_not_end = meta.selector();
+        let q_word_end = meta.complex_selector();
+        let r_multi = meta.advice_column();
+        let input = meta.advice_column();
+        let data = meta.advice_column();
+        let data_rlc = meta.advice_column();
+        let num_bits = meta.advice_column();
+        let shift_factor = meta.advice_column();
+        let decode_factor = meta.advice_column();
+        let huffman_table = array_init::array_init(|_| meta.lookup_table_column());
+        let fixed_table = array_init::array_init(|_| meta.lookup_table_column());
+
+        meta.enable_equality(data_rlc);
+
+        // range check
+        // input(0..256)
+        // data(0..256)
+        // num_bits(0..32)
+        // r_multi(bool)
+        // shift_factor 2.pow(0..8)
+        // decode_factor 2.pow(0..32)
+        macro_rules! range_check {
+            ($column:ident, $range:ident) => {
+                meta.lookup(concat!(stringify!($column), " range check"), |meta| {
+                    let column = meta.query_advice($column, Rotation::cur());
+                    let q_enable = meta.query_selector(q_enable);
+                    vec![
+                        (
+                            q_enable.clone() * FixedTableTag::$range.expr(),
+                            fixed_table[0],
+                        ),
+                        (q_enable * column, fixed_table[1]),
+                    ]
+                });
+            };
+        }
+
+        range_check!(input, Byte);
+        range_check!(data, Byte);
+        range_check!(num_bits, Range32);
+        range_check!(r_multi, Bool);
+        range_check!(shift_factor, Exponent8);
+        range_check!(decode_factor, Exponent32);
```

Comment:
Hmmm so not sure I understand why you do it like this. Shouldn't this be a single lookup into a specifically created table? Because this will simply check that each value separately is in the correct range, but this will not check that all values together are the correct values to use for the current state no? Or how are you planning on checking that these values are the ones necessary (e.g. 5 bits are already used, we decode a word with 2 bits so we need to add these 2 bits to the data stream multiplied by 2. These values are all interconnected and so if you put them all in a single table and lookup all values together you know all of them are correctly set and in the expected range, using a single lookup).

Reply:
Yeah, you're right. So I need check-range through all parts of one word

Reply:
Because I have check the data(decode_factor)/num_bits in "lookup huffman table". So I only needs to check the input/r_multi/shift_factor

Reply:
I still don't really understand. For example `shift_factor`, how do we verify it's the correct value in the circuit? Using my example above:

> 5 bits are already used, we decode a word with 2 bits so we need to add these 2 bits to the data stream multiplied by 2. 

How does the circuit verify that `shift_factor` needs to be `2` in this case? Any other value would not create the correct rlc encoding of the data stream. It seems to me the prover could currently supply e.g. `1` which should not be valid.